### DISCUSSION TITLE
Rapidchange Automatic tool changer, can it be made to work with UGS?

### DISCUSSION THREAD

**Discussion by jappiemike** [winder/Universal-G-Code-Sender#2257] [Help]
Hello, I've come across this automatic tool changer which is compatible with the Fluidnc board I'm using on my gantry CNC. I asked the question on their discord site if UGS was compatible and when they checked it seems to complete the file but throws some errors. I'm wondering if this is a fatal in the long run or something I/we can work around. The video Don Greilick, the creator of this ATC, is too big to attach so here is a link to the page https://discord.com/channels/1077383334080032898/1106933931938418779.
Thank you for all the work you've all put in to this fantastic piece of software.
Cheers
Mike

**Comment by breiler** [comment]
The link you posted did not work, but I found their channel here: https://www.youtube.com/@RapidChangeCNC
And website: https://rapidchangecnc.com/

As I understand it, this is handled completely by the controller and is transparent to the sender. 
What errors are thrown?

**Comment by jappiemike** [comment]
The error is "Error while processing response <ok> : An unexpected command was completed by the contoller

https://github.com/winder/Universal-G-Code-Sender/assets/892453/4cbe7720-6edc-45da-8e4c-ef2169e19369

I've managed to trim the file to attach it.

**Comment by breiler** [comment]
Thanks for the video.

The errors means that UGS receives an "ok" response for a command that it didn't send. There is a risk that UGS will end up out of sync thinking that the controller has `ok`:ed sent commands.

I think that their implementation of this is a bit wrong. In traditional GRBL you send a command to the controller and it responds with either an `ok` or `error`. It seems that they have implemented this so that "M6T1" expands to multiple commands which gets `ok`:ed and returned to the UGS. In my opinion there should only be one ok from the controller once the tool change operation is finished.

**Reply by jappiemike** [reply]
Thank's for your time in looking into what is essentially not your problem. I've sent your reply over to the folks at RapidChange and will keep you in the loop.
Cheers
Mike
